src/metric/auc.cu

```diff
@@ -0,0 +1,547 @@
+struct DeviceAUCCache {
+  // Pair of FP/TP
+  using Pair = thrust::pair<float, float>;
+  // index sorted by prediction value
+  dh::device_vector<size_t> sorted_idx;
+  // track FP/TP for computation on trapesoid area
+  dh::device_vector<Pair> fptp;
+  // track negative/positive for computation on trapesoid area
+  dh::device_vector<Pair> neg_pos;
+  // index of unique prediction values.
+  dh::device_vector<size_t> unique_idx;
+  // p^T: transposed prediction matrix, used by MultiClassAUC
+  dh::device_vector<float> predts_t;
+  std::unique_ptr<dh::AllReducer> reducer;
+
+  void Init(common::Span<float const> predts, bool is_multi, int32_t device) {
+    if (sorted_idx.size() != predts.size()) {
+      sorted_idx.resize(predts.size());
+      fptp.resize(sorted_idx.size());
+      unique_idx.resize(sorted_idx.size());
+      neg_pos.resize(sorted_idx.size());
+      if (is_multi) {
+        predts_t.resize(sorted_idx.size());
+        reducer.reset(new dh::AllReducer);
+        reducer->Init(rabit::GetRank());
+      }
+    }
+  }
+};
+
+/**
+ * The GPU implementation uses same calculation as CPU with a few more steps to distribute
+ * work across threads:
+ *
+ * - Run scan to obtain TP/FP values, which are right coordinates of trapesoid.
+ * - Find distinct prediction values and get the corresponding negative/postive value,
+ *   which are left coordinates of trapesoid.
+ * - Reduce the scan array into 1 AUC value.
+ */
+std::tuple<float, float, float>
+GPUBinaryAUC(common::Span<float const> predts, MetaInfo const &info,
+             int32_t device, std::shared_ptr<DeviceAUCCache> *p_cache) {
+  auto& cache = *p_cache;
+  if (!cache) {
+    cache.reset(new DeviceAUCCache);
+  }
+  cache->Init(predts, true, device);
+
+  auto labels = info.labels_.ConstDeviceSpan();
+  auto weights = info.weights_.ConstDeviceSpan();
+  dh::safe_cuda(cudaSetDevice(device));
+
+  CHECK(!labels.empty());
+  CHECK_EQ(labels.size(), predts.size());
+
+  /**
+   * Create sorted index for each class
+   */
+  auto d_sorted_idx = dh::ToSpan(cache->sorted_idx);
+  dh::ArgSort<false>(predts, d_sorted_idx);
+
+  /**
+   * Linear scan
+   */
+  dh::caching_device_vector<float> d_auc(1);
+  auto s_d_auc = dh::ToSpan(d_auc);
+  dh::LaunchN(device, d_auc.size(),
+              [=] __device__(size_t i) { s_d_auc[i] = 0; });
+
+  auto get_weight = GetWeightOp{weights, d_sorted_idx};
+  using Pair = thrust::pair<float, float>;
+  auto get_fp_tp = [=]__device__(size_t i) {
+    size_t idx = d_sorted_idx[i];
+
+    float label = labels[idx];
+    float w = get_weight(i);
+
+    float fp = (1.0 - label) * w;
+    float tp = label * w;
+
+    return thrust::make_pair(fp, tp);
+  };  // NOLINT
+  auto d_fptp = dh::ToSpan(cache->fptp);
+  dh::LaunchN(device, d_sorted_idx.size(),
+              [=] __device__(size_t i) { d_fptp[i] = get_fp_tp(i); });
+
+  dh::XGBDeviceAllocator<char> alloc;
+  auto d_unique_idx = dh::ToSpan(cache->unique_idx);
+  dh::Iota(d_unique_idx, device);
```

Comment:
Is an Iota function necessary? What about thrust sequence?

Reply:
Written for the `ArgSort`, which is async.